### ISSUE TITLE
Add: Destroying buildings damages your reputation

### ISSUE THREAD

**Issue by mtusnio** [Vdauphin/HeartsAndMinds#288]
I noticed that with 1.68 a new mission event handler will arrive called "BuildingChanged". I was hoping to expand on the reputation system a bit for my group, but I'm not sure how much of it would get accepted up here in the upstream. The general idea is to use BuildingChanged to track whenever a building is destroyed (regardless of the source - it doesn't look like it'll be easy to track that, plus the locals don't care, they hate you for whatever the rebels do by accident) and take away reputation for that. I want also to put in a filter so that more important buildings (churches, supermarkets, hotels) count more against your reputation than wooden sheds.

As a bonus I was hoping to find a way to add saving of destroyed buildings using this handler as well. My group blew up a church lately to clear an ammo cache (we forgot we can just take it out of the church),  it'd be a nice touch to keep it destroyed between saves. :)

Currently blocked on 1.68

**Comment by Vdauphin** [comment]
> I noticed that with 1.68 a new mission event handler will arrive called "BuildingChanged".

Interesting! 

https://community.bistudio.com/wiki/Arma_3:_Event_Handlers/addMissionEventHandler#BuildingChanged

> each time a building model changes

We need to be sure that this EH is not to much fired. What does it mean "model changes" ?

> (regardless of the source - it doesn't look like it'll be easy to track that, plus the locals don't care, they hate you for whatever the rebels do by accident)

Agree, Enemy are not destroying a lot of building, I think. Is it wrong?

>  I want also to put in a filter so that more important buildings (churches, supermarkets, hotels) count more against your reputation than wooden sheds.

Why not, sound cool.

> As a bonus I was hoping to find a way to add saving of destroyed buildings using this handler as well. My group blew up a church lately to clear an ammo cache (we forgot we can just take it out of the church), it'd be a nice touch to keep it destroyed between saves. :)

I will be agree if it is optional. Most of the time server lag come from destroyed building. 

(use master_daily branch for this one :) )

**Comment by Breech99** [comment]
I love this idea of prioritized VIP structures like churches, schools and other that will punish blufor if destroyed. Keep from indiscriminate leveling of villages. Surgical strikes from laser guided munitions and clearing out areas with foot mobiles.

**Comment by mtusnio** [comment]
@Vdauphin 

> We need to be sure that this EH is not to much fired. What does it mean "model changes" ?

From what I understand it does it whenever it changes due to different stages of destruction, so unless you level an entire city in one go it shouldn't be that big of a deal. I'll do some proper testing with 1.68 to make sure it doesn't cause any issues. :)

> Agree, Enemy are not destroying a lot of building, I think. Is it wrong?

Sometimes they might miss with an RPG, I was more thinking of an odd IED blowing buildings up. I imagined though that narrative wise this means that the locals just don't know what happened or why, so they blame you anyway. :D 

> I will be agree if it is optional. Most of the time server lag come from destroyed building.

Good point, we play really small scale missions, so most of the time it's not an issue. I'll open a new ticket for this once I start working on it, don't want to bundle up all too many features under this one.

@Breech99 

> I love this idea of prioritized VIP structures like churches, schools and other that will punish blufor if destroyed. Keep from indiscriminate leveling of villages. Surgical strikes from laser guided munitions and clearing out areas with foot mobiles. Well as long as it doesn't impact server performance.

Exactly, generally I really like the reputation system and I'm looking into ways to expand the bonuses/maluses and see if there's a way to add more civilian interaction. :) 

> As a side note: I am not sure if cba or what specifically is causing the 3fps bug on some machines. I first thought it was the scripts in @CBA mod but now it looks as it's a visual graphic memory leak or something else. We hope BIS finds it and squashes the problem.

AFAIK this is fixed in 1.68, the issue was that the 32-bit client could only address about 4 GB of memory and sometimes would run into some weird out-of-memory state. The 64-bit client should fix that!

In other news, apparently 1.68 is coming possibly this week! https://dev.arma3.com/post/sitrep-00186 I might wait for that before any development, don't want to keep switching between dev & release builds.

**Comment by Vdauphin** [comment]
> From what I understand it does it whenever it changes due to different stages of destruction, so unless you level an entire city in one go it shouldn't be that big of a deal. I'll do some proper testing with 1.68 to make sure it doesn't cause any issues. :)

Understood :)


> I was more thinking of an odd IED blowing buildings up. 

Good point. If IED blow, it is because a player trigger it. So it is normal that player get malus if he trigger the IED.

> In other news, apparently 1.68 is coming possibly this week! https://dev.arma3.com/post/sitrep-00186 I might wait for that before any development, don't want to keep switching between dev & release builds.

Perfect!

**Comment by mtusnio** [comment]
And 1.68 is officially out, yay!

**Comment by Breech99** [comment]
Yes. So smooth. Now depending on the various state changes in buildings and depending on how often the event handler has to fire off, I think it would be best to just give negative reputation for buildings destroyed. I can dig up the class names if interested. Possible list would be small churches, large church, red/white hotel, supermarkets. Not sure how to have it done via map other than possibly a map placed Game logic to locate the proximity of the building that will use the event handlers? buildings nearby the game logic?

**Comment by mtusnio** [comment]
I don't think you need anything placed on the map, I imagine the event handler should take care of everything for you. There might be some data structure required for saving them, but that's going to be a different issue entirely. :)

**Comment by Breech99** [comment]
Now that this command is in 1.68, how are we going to apply it? I've read it's pretty simple it's just how  it's going to be used like I would say only for destruction of a target building and only a few of the types of buildings.  Awesome idea, let's make this happen. It's good to be careful and not level a whole village heh.

**Comment by mtusnio** [comment]
Sorry for the delay, I've been quite distracted lately. I'm hoping to do some basic implementation this weekend if possible, but it all depends whether I'm actually at my PC. :) The basic idea is to just add new mission handler and subtract reputation then. Then check how often it's called, whether it covers edge cases and whether the "isRuin" flag works as you'd expect, since that could be used to add extra penalty for fully destroying a building. Then it's just a case of adding an array which gives you the mapping from building string to the modifier. That might be slightly tricky if the string changes drastically between stages, but I guess then you can build a dynamic translation array on the fly.

**Comment by Vdauphin** [comment]
Hello, nothing really complex ?

```
addMissionEventHandler ["BuildingChanged",{
	params ["_change_from","_change_to","_isruin"];
	if (_isruin) then {btc_rep_malus_civ_house call btc_fnc_rep_change;};
}];
```

**Comment by Breech99** [comment]
I tossed the idea to the forums to see what kind of  feedback we got.  Here's a cool idea presented by WyattWic on the bohemia forums:

The issue with that EH is it returns on every building destroy phase.  Here is my workaround for one of my database scripts.
 
On initServer.sqf
addMissionEventHandler ["BuildingChanged",{[_this select 0,_this select 1,_this select 2] call db_fnc_mapbuildings;}]; //Handles db_fnc_mapbuildings
On db_fnc_mapbuildings (or whatever name you give it in your functions libary)
if (isnil "MapDamage") then {
	MapDamage = [];			
}; //Create our MD variable if we haven't already.

private _object = _this select 0;
private _newobject = _this select 1;
private _ruin = _this select 2;
private _netID = netid _object;

if (_ruin) then {
//It's completely destroyed, based on the return from the mission event handler.

	//If its not a native map object, see if it's a product of a previously recorded destruction.
	if ((netid _object find "-") == -1) then { //Not a native.  
		if !((MapDamage select {_x select 1 == netid _object}) isEqualTo []) then {
			_oldobject = ((MapDamage select {_x select 1 == netid _object}) select 0) select 0;//This should return the netID of its original object
			
			//This code here will run if the item passed was not native, but was a product of a native object decaying into ruin.
			
		} else {
			
			//This code here will run if it was not a map object and not a product of a map object decaying into ruin.
			//AKA a script placed item dieing completely.
			//XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
			
		};
	} else { 
		//It is a native object being placed as ruined.
	};
} else {
//It's only partially damaged.
	//If its not a native map object, see if it's a product of a previously recorded destruction.
	if ((netid _object find "-") == -1) then { 
		if !((MapDamage select {_x select 1 == netid _object}) isEqualTo []) then {
			_oldobject = str (MapDamage select {_x select 1 == netid _object}) select 0;
			
			//Its a native map object passing through another phase of its destruction, but is not ruined yet.
			
		} else {
		
			//It was never a native map object and it's not completely ruined.
			//XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
		};
	} else {
		//A native map object produced a non-native object.
		MapDamage pushBack [netid _object,netid _newobject];

	};
};
I've tested it lightly, but it hasn't seen much usage yet.  The sections your interested in have a bunch of Xes in it.
Even though your not interested in half-dead structures, the whole thing is needed to keep track of what are byproducts of the native map structures.
 
Hope that helps!

**Comment by Breech99** [comment]
Your simple way might be easier too dauphin.  Just like you made for the fired near scripts. Have an  addEventHandler and a RemoveEventHandler to prevent spamming.

**Comment by mtusnio** [comment]
Hey Vdauphin,
I added someone with your nickname on Steam (my profile - http://steamcommunity.com/id/maverick001 ), it might be easier to fill you in there. :)

**Comment by Vdauphin** [comment]
Sorry, I have a lot of request on steam. I just added you :)
